### ISSUE TITLE
MachinePool reporting `Ready` as true if infrastructure ref isn't found

### ISSUE THREAD

**Issue by richardcase** [kubernetes-sigs/cluster-api#4311]
**What steps did you take and what happened:**
I have a typo for the **infrastructureRef** on a **MachinePool** so:

1. Created a `MachinePool` using an infrastructure ref for CAPA:

```
apiVersion: exp.cluster.x-k8s.io/v1alpha3
kind: MachinePool
metadata:
  name: "wkp-mgmt01-pool-0"
spec:
  clusterName: "wkp-mgmt01"
  replicas: 2
  failureDomains:
    - eu-west-2c
  template:
    spec:
      clusterName: "wkp-mgmt01"
      bootstrap:
        dataSecretName: ""
      infrastructureRef:
        name: "wpk-mgmt01-pool-0"
        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
        kind: AWSManagedMachinePoolyaml
```
2.  Due to a typo the infrastructure ref name (`wpk-mgmt01-pool-0`) is incorrect and there is no match
3. If i have a wait on the `Ready` condition of the `MachinePool` is showing ready is true:
![image](https://user-images.githubusercontent.com/198425/111148068-1b4a5b80-8583-11eb-8fb6-2d71d81b7cd2.png)
4. And then if i describe the MachinePool you set the ready condition is true and the phase is provisioning:
![image](https://user-images.githubusercontent.com/198425/111148206-43d25580-8583-11eb-848f-22073d90b2b2.png)

**What did you expect to happen:**
I would expect the ready condition on the MachinePool to be false and an event written indicating that infrastructure ref can't be found.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api version: 0.3.14
- Minikube/KIND version: 0.10.0
- Kubernetes version: (use `kubectl version`): 1.18
- OS (e.g. from `/etc/os-release`): Linux

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

**Comment by srm09** [comment]
Was able to reproduce this issue locally with the docker machinepool as well. Looking into it.

**Comment by srm09** [comment]
This happens since if we have a failure to reconcile the externalRef, the reconcile loop returns an error without setting the `InfrastructureReady` condition on the MachinePool. I expect a similar behavior for the bootstrapRef too.

Looking at other controllers, this could be the case with `Machine` objects as well.

**Comment by fabriziopandini** [comment]
> if any of the participating conditions is not set/found, we should set the Ready condition as false (instead of the default true)

This could be problematic, because we don't have any guarantee/any raccomandation in current condition spec that all the conditions should be created on an object (might be they appear sometime later in the processing, like e.g for KCP upgrades, or might be there are conditions which  appears only in case of specific topologies configurations).

instead, if there is a problem managing an external reference, IMO we should set the condition representing the external ref (e.g InfrastructureReadyCondition or BootstrapReadyCondition on Machines) to false and surface the issue.

**Comment by srm09** [comment]
The solution you are proposing is a simple resolution and makes total sense. But since, the Ready condition is an aggregate/summation of a list of specified conditions, if one of them does not exist/or is yet top be reported, setting the condition to be Ready is a false view of the actual state of the object.

For example, in case of KCP upgrades, if the control plane is undergoing a rolling upgrade, shouldn't it be marked as **not** `Ready` until all of the replicas are upgraded. This is the case that am trying to make, that the Ready condition should always reflect the operational state of the object.

For cases in which certain conditions appear only in specific topologies, we could have an indicator while generating the condition which states whether the condition should be present and taken into account while constructing the `Ready` condition.

**Comment by fabriziopandini** [comment]
I think we should break the discussion in two part
1. this specific issue, where it seems we agree that the solution is to make InfrastructureReadyCondition or BootstrapReadyCondition to surface problems in reconciling external resources
2. how summary should work, and this IMO should be tracked as a separated issue. 

For the sake of record, WRT to 2 I'm not opposed to the change in principle, but this is not the current assumption and so, before choosing another approach we should make some more investigation about impacts, not only in CAPI.  
BTW, as of today, for upgrades there is a condition that surface there is an upgrade in progress, and this makes KCP to be not ready during the entire upgrade, so we are already reflecting the operational state of the object even with the current summary logic

**Comment by srm09** [comment]
That makes sense, I will create a new issue to track the point 2 that we were discussing. As for this issue, I will move forward with setting the state of the condition for external references for MachinePools and create other issues for other controllers.
